File: examples/typescript/src/gcs_context.ts
import type { CommandProtocol } from '../../../generated/ts/protocols/command_protocol.ts';
import type {
  HeartbeatMonitor,
  HeartbeatPublisher,
  MavlinkNode,
} from '../../../generated/ts/protocols/heartbeat_protocol.ts';
import type { MavlinkCancellationToken } from '../../../generated/ts/protocols/mavlink_cancellation.ts';
import type { MavlinkSession } from '../../../generated/ts/protocols/mavlink_session.ts';
import type {
  MavlinkGcs,
  MavlinkVehicleClient,
} from '../../../generated/ts/protocols/mavlink_vehicle_client.ts';
import type { MissionProtocol } from '../../../generated/ts/protocols/mission_protocol.ts';
import type { ParameterProtocol } from '../../../generated/ts/protocols/parameter_protocol.ts';

/** Ground control station identity (MAVLink convention). */
export const gcsSystemId = 255;
export const gcsComponentId = 190;

/** Shared MAVLink GCS state for the interactive SITL example. */
export class GcsContext {
  readonly gcs: MavlinkGcs;
  readonly vehicle: MavlinkNode;
  readonly client: MavlinkVehicleClient;

  /** Cancels in-flight parameter/mission operations (type `cancel` in CLI). */
  operationCancel: MavlinkCancellationToken | null = null;

  constructor(options: {
    gcs: MavlinkGcs;
    vehicle: MavlinkNode;
    client: MavlinkVehicleClient;
  }) {
    this.gcs = options.gcs;
    this.vehicle = options.vehicle;
    this.client = options.client;
  }

  get session(): MavlinkSession {
    return this.gcs.session;
  }

  get heartbeatMonitor(): HeartbeatMonitor {
    return this.gcs.heartbeatMonitor;
  }

  get heartbeatPublisher(): HeartbeatPublisher {
    return this.gcs.heartbeatPublisher;
  }

  get parameters(): ParameterProtocol {
    return this.client.parameters;
  }

  get mission(): MissionProtocol {
    return this.client.mission;
  }

  get command(): CommandProtocol {
    return this.client.command;
  }

  get targetSystem(): number {
    return this.vehicle.systemId;
  }

  get targetComponent(): number {
    return this.vehicle.componentId;
  }
}

File: examples/rust/src/gcs_context.rs
//! Shared MAVLink GCS state for the interactive SITL example.

use mavlink::protocols::{
    CommandProtocol, HeartbeatMonitor, HeartbeatPublisher, MavlinkCancellationToken,
    MavlinkGcs, MavlinkNode, MavlinkSession, MavlinkVehicleClient, MissionProtocol,
    ParameterProtocol,
};

/// Ground control station identity (MAVLink convention).
pub const GCS_SYSTEM_ID: u8 = 255;
pub const GCS_COMPONENT_ID: u8 = 190;

/// Shared MAVLink GCS state for the interactive SITL example.
pub struct GcsContext {
    pub gcs: MavlinkGcs,
    pub vehicle: MavlinkNode,
    pub client: MavlinkVehicleClient,
    /// Cancels in-flight parameter/mission operations (`cancel` CLI command).
    pub operation_cancel: Option<MavlinkCancellationToken>,
}

impl GcsContext {
    pub fn new(gcs: MavlinkGcs, vehicle: MavlinkNode, client: MavlinkVehicleClient) -> Self {
        Self {
            gcs,
            vehicle,
            client,
            operation_cancel: None,
        }
    }

    pub fn session(&self) -> &MavlinkSession {
        &self.gcs.session
    }

    pub fn heartbeat_monitor(&self) -> &HeartbeatMonitor {
        &self.gcs.heartbeat_monitor
    }

    pub fn heartbeat_publisher(&self) -> &HeartbeatPublisher {
        &self.gcs.heartbeat_publisher
    }

    pub fn parameters(&self) -> &ParameterProtocol {
        &self.client.parameters
    }

    pub fn mission(&self) -> &MissionProtocol {
        &self.client.mission
    }

    pub fn command(&self) -> &CommandProtocol {
        &self.client.command
    }

    pub fn target_system(&self) -> u8 {
        self.vehicle.system_id
    }

    pub fn target_component(&self) -> u8 {
        self.vehicle.component_id
    }
}
